test/python/circuit/test_hamiltonian_gate.py

```diff
@@ -0,0 +1,203 @@
+
+    def test_transpose(self):
+        """test transpose"""
+        ham = HamiltonianGate([[15, 1j], [-1j, -2]], np.pi / 7)
+        np.testing.assert_array_almost_equal(ham.transpose().to_matrix(),
+                                             np.transpose(ham.to_matrix()))
+
+    def test_adjoint(self):
+        """test adjoint operation"""
+        ham = HamiltonianGate([[3, 4j], [-4j, -.2]], np.pi * .143)
+        np.testing.assert_array_almost_equal(ham.adjoint().to_matrix(),
+                                             np.transpose(np.conj(ham.to_matrix())))
+
+
+class TestHamiltonianCircuit(QiskitTestCase):
+    """Hamiltonian gate circuit tests."""
+
+    def test_1q_hamiltonian(self):
+        """test 1 qubit hamiltonian"""
+        qr = QuantumRegister(1, 'q0')
+        cr = ClassicalRegister(1, 'c0')
+        qc = QuantumCircuit(qr, cr)
+        matrix = np.zeros((2, 2))
+        qc.x(qr[0])
+        theta = Parameter('theta')
+        qc.append(HamiltonianGate(matrix, theta), [qr[0]])
+        qc = qc.bind_parameters({theta: 1})
+
+        # test of text drawer
+        self.log.info(qc)
+        dag = circuit_to_dag(qc)
+        dag_nodes = dag.named_nodes('hamiltonian')
+        self.assertTrue(len(dag_nodes) == 1)
+        dnode = dag_nodes[0]
+        self.assertIsInstance(dnode.op, HamiltonianGate)
+        self.assertListEqual(dnode.qargs, qc.qubits)
+        assert_allclose(dnode.op.to_matrix(), np.eye(2))
+
+    def test_error_on_qasm(self):
+        """test that an error is thrown if qc.qasm() is called. """
+        qr = QuantumRegister(1, 'q0')
+        cr = ClassicalRegister(1, 'c0')
+        qc = QuantumCircuit(qr, cr)
+        matrix = np.zeros((2, 2))
+        qc.hamiltonian(operator=matrix, time=1, qubits=qr[0])
+
+        with self.assertRaises(ExtensionError):
+            qc.qasm()
+
+    def test_2q_hamiltonian(self):
+        """test 2 qubit hamiltonian """
+        qr = QuantumRegister(2)
+        cr = ClassicalRegister(2)
+        qc = QuantumCircuit(qr, cr)
+        matrix = Operator.from_label('XY')
+        qc.x(qr[0])
+        theta = Parameter('theta')
+        uni2q = HamiltonianGate(matrix, theta)
+        qc.append(uni2q, [qr[0], qr[1]])
+        passman = PassManager()
+        passman.append(CXCancellation())
+        qc2 = passman.run(qc)
+
+        # Testing bind after transpile
+        qc2 = qc2.bind_parameters({theta: -np.pi / 2})
+        # test of text drawer
+        self.log.info(qc2)
+        dag = circuit_to_dag(qc2)
+        nodes = dag.two_qubit_ops()
+        self.assertEqual(len(nodes), 1)
+        dnode = nodes[0]
+        self.assertIsInstance(dnode.op, HamiltonianGate)
+        for qubit in dnode.qargs:
+            self.assertIn(qubit.index, [0, 1])
+        # Equality based on Pauli exponential identity
+        np.testing.assert_array_almost_equal(dnode.op.to_matrix(), 1j*matrix.data)
+        qc3 = dag_to_circuit(dag)
+        self.assertEqual(qc2, qc3)
```

Comment:
Not blocking, but the tests would be easier to follow if they were each a bit narrower in scope. Ideally, covering a single property or use-case per test.

Reply:
I tried to copy the Unitary tests as closely as possible, but I get your point, I've tried to remove some of the clutter.